From 145416c70b2e4422f0ff03f402da33a4a4db29e2 Mon Sep 17 00:00:00 2001
From: Jeremy Harris <jgh146exb@wizmail.org>
Date: Sat, 29 Jun 2019 19:37:57 +0100
Subject: [PATCH 29/29] Fix bounce generation under RFC 3461 request.  Bug 2411

Broken-by: ea97267cea
(cherry picked from commit df98a6ff2e70887890690ffbf8a8ad583d7d7e38)
(cherry picked from commit b4a37a77271a8f6efc887d68265eb7867eff6170)
---
 doc/ChangeLog        |  4 ++++
 src/deliver.c            |  3 ++-
 test/log/0618                | 29 +++++++++++++++++++++++++++
 test/rejectlog/0618          |  1 +
 test/scripts/0000-Basic/0618 | 38 ++++++++++++++++++++++++++++++++++++
 test/stdout/0618             | 31 +++++++++++++++++++++++++++++
 6 files changed, 105 insertions(+), 1 deletion(-)

diff --git a/doc/ChangeLog b/doc/ChangeLog
index 93d3349b..e92ca240 100644
--- a/doc/ChangeLog
+++ b/doc/ChangeLog
@@ -82,6 +82,10 @@ JH/29 Fix DSN Final-Recipient: field.  Previously it was the post-routing
       forwarding.  Change to the original envelope recipient address, per
       standards.
 
+JH/30 Bug 2411: Fix DSN generation when RFC 3461 failure notification is
+      requested.  Previously not bounce was generated and a log entry of
+      error ignored was made.
+
 
 Exim version 4.92
 -----------------
diff --git a/src/deliver.c b/src/deliver.c
index 53562dd5..d4ed8af0 100644
--- a/src/deliver.c
+++ b/src/deliver.c
@@ -7519,7 +7519,8 @@ while (addr_failed)
   mark the recipient done. */
 
   if (  addr_failed->prop.ignore_error
-     || addr_failed->dsn_flags & (rf_dsnflags & ~rf_notify_failure)
+     ||    addr_failed->dsn_flags & rf_dsnflags
+	&& !(addr_failed->dsn_flags & rf_notify_failure)
      )
     {
     addr = addr_failed;
-- 
2.20.1

